[nnc] Merge inconsistent profiling information

bertmaher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#60510 [nnc] Merge inconsistent profiling information**

We encountered a situation where loop unrolling caused us to duplicate
profiled tensor types in a manner that wasn't logically consistent (see the
attached test case).  When applying this profiling information, we need to
merge the profiled types so that we use a conservative (unspecialized) type.

Differential Revision: [D29322487](https://our.internmc.facebook.com/intern/diff/D29322487/)

bertmaher committed:
[nnc] Merge inconsistent profiling information

We encountered a situation where loop unrolling caused us to duplicate
profiled tensor types in a manner that wasn't logically consistent (see the
attached test case).  When applying this profiling information, we need to
merge the profiled types so that we use a conservative (unspecialized) type.

Differential Revision: [D29322487](https://our.internmc.facebook.com/intern/diff/D29322487/)

[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/passes/tensorexpr_fuser.cpp:
```diff
@@ -313,7 +313,12 @@ void removeProfileNodesAndSpecializeTypes(Block* b) {
+      // them.  This situation can happen if, e.g., loop unrolling duplicates
+      // profiled types in a loop body in a manner that isn't logically
+      // consistent (see TestTEFuser.test_unrolled_cat).
+      it->input()->setType(
+          it->input()->type()->expect<TensorType>()->merge(*profiled_type));
```
Isn't this always resulting in non-specified type? In what situations `it->input()->type()->expect<TensorType>()` would have a concrete type?

ZolotukhinM replied:
I think with this change the fuser might just end up fusing nothing

Krovatkin replied:
if two types are concrete and match exactly, we will get the complete type.
We would probably end up fusing less for sure. That's pretty much a main reason why this logic was implemented the way it is i.e. we wanted to speculate and pick a concrete type.

bertmaher replied:
Oh yeah, this basically disables the fuser.  I mistakenly thought it->input() actually had a type, but it does not, it's always Tensor (except for those rare cases where it is simply wrong).

Krovatkin replied:
@bertmaher it should probably be : 

```cpp
      if (profiled_type == TensorType::get()) {
        continue;
      }

      if (it->input()->type() == TensorType::get()) {
          it->input()->setType(it->ty(attr::profiled_type));
      }
      else {
          it->input()->setType(it->input()->merge(it->ty(attr::profiled_type)));
      }
```

navahgar replied:
> Oh yeah, this basically disables the fuser. I mistakenly thought it->input() actually had a type, but it does not, it's always Tensor (except for those rare cases where it is simply wrong).

This is exactly what I pointed out in the chat earlier today. Nick's latest fix should handle this case.

Krovatkin replied:
if you guys really don't trust merge and I can't blame you,
we could try something like this?

```cpp
      if (it->input()->type() == TensorType::get() || *it->ty(attr::profiled_type) == *it->input()->type()) {
          it->input()->setType(it->ty(attr::profiled_type));
      }
```

Note `*` otherwise we would be comparing pointers to types rather then calling the override for ==. 
`it->input()->type() == TensorType::get()` is okay though since there could be only one `TensorType::get()`

ZolotukhinM reviewed (commented):
Wouldn't a cleaner fix be to handle profile nodes in loop unroller (and not duplicate them in duplicated iterations)?

bertmaher commented:
Yeah this basically just turns off the fuser, it does not do what I thought it did.

bertmaher commented:
> Wouldn't a cleaner fix be to handle profile nodes in loop unroller (and not duplicate them in duplicated iterations)?

I'm not sure.  I kind of think sometimes you might want duplicated nodes, or you might miss fusion opportunities.  I kind of liked this approach because it was "generic" in that it finds anything with conflicting profile data and reconciles it.

Krovatkin commented:
> I'm not sure. I kind of think sometimes you might want duplicated nodes, or you might miss fusion opportunities. I kind of liked this approach because it was "generic" in that it finds anything with conflicting profile data and reconciles it.

I'm somewhat split. 
I also think we should come up with a cleaner fix down the road. 
My concern with the loop fix though is that this exact issue may crop up with in-place ops as well. I'm trying to come up with a small example. The only difference is that we wouldn't try fusing inplace ops, so the issue may lurk for ahwile.

bertmaher commented:
> I also think we should come up with a cleaner fix down the road.

+1, this fix (even the corrected version you provided) still doesn't give me warm fuzzy feelings.  I think the best approach is to not treat profiling data as ground truth, and instead use shape inference to figure out what type the fusion group should actually produce.

navahgar commented:
@Krovatkin 
> we wouldn't try fusing inplace ops
Currently, yes. But we do have plans to handle them in H2.

Krovatkin commented:
> not treat profiling data as ground truth, and instead use shape inference to figure out what type the fusion group should actually produce.
+100 this 

Profiling assumes that we check the properties right before the operation.
However, here we test the properties of some inputs up the chain and hope that the profiling info on the results is consistent.

bertmaher committed:
Update on "[nnc] Merge inconsistent profiling information"


We encountered a situation where loop unrolling caused us to duplicate
profiled tensor types in a manner that wasn't logically consistent (see the
attached test case).  When applying this profiling information, we need to
merge the profiled types so that we use a conservative (unspecialized) type.

Differential Revision: [D29322487](https://our.internmc.facebook.com/intern/diff/D29322487/)

[ghstack-poisoned]

bertmaher commented:
Okay, sorry, I am officially slower than @navahgar, @ZolotukhinM, and @Krovatkin.  4/4 = 100% right, so I'm good?  I blame actually spending the day fighting with the repro case itself :-).  Anyways this corrected version should actually fix the problem without disabling the fuser entirely.

Krovatkin reviewed (approved):
👍

ZolotukhinM commented on torch/csrc/jit/passes/tensorexpr_fuser.cpp:
```diff
@@ -313,7 +313,16 @@ void removeProfileNodesAndSpecializeTypes(Block* b) {
+      // profiled types in a loop body in a manner that isn't logically
+      // consistent (see TestTEFuser.test_unrolled_cat).
+      auto input_type = it->input()->type()->expect<TensorType>();
+      if (input_type == TensorType::get()) {
+        it->input()->setType(profiled_type);
+      } else {
+        it->input()->setType(input_type->merge(*profiled_type));
+      }
```
I still think the fix is invalid: e.g. this check assumes there would be conflicting types in the graph - but in the test all the types in the graph are either "Float(1)" or "Tensor" which means that we're either disabling the fuser entirely or not catching anything. I believe at this point it is too late(*) to try to find inaccuracies in profiling data - it should be done when that data is introduced and that happens in two places currently: 1) during profiling itself, 2) in loop unroller when it copies profiling data for the subsequent iterations. I'm sure (1) handles conflicting types correctly, but (2) can't do that and hence I think it should just drop the data.

\* _We still can try finding inconsistencies in the profiling data if we employ an alternative mechanism to obtain that info - e.g. shape inference. But without an alternative there is simply nothing to verify against._

ZolotukhinM replied:
If you insert `assert(false)` into the else branch, will it ever fire?
